### DISCUSSION TITLE
Using DirectWrite on Windows

### DISCUSSION THREAD

**Discussion by dgregorius** [ocornut/imgui#5864] [New Users Build/Link/Run issues ONLY!]
Did anybody think about this? I wonder if this a feasible solution to improve font rendering in ImGui (on Windows). I am wondering if one could use DirectWrite with the DirectX11 and DirectX12 backends. Or even one step further- did anybody even try to implement a full Direct2D/DirectWrite backend?

An alternative would be to use DirectWrite to create the font atlas. If this is possible it might be a feasible solution.

Just musing. Sorry if this doesn't make sense!

**Comment by PathogenDavid** [comment]
See https://github.com/ocornut/imgui/issues/3365 and Omar's response in https://github.com/ocornut/imgui/issues/3406.

> An alternative would be to use DirectWrite to create the font atlas. If this is possible it might be a feasible solution.

This should be possible, but is DirectWrite's font rendering any better than what you get with [imgui_freetype](https://github.com/ocornut/imgui/tree/44d98bfd6d12caee43bea49fda71c8313c3e57e7/misc/freetype)?

I don't think the rendering of individual glyphs is the main issue with Dear ImGui's text, it's the lack of a dynamic font atlas and missing sophisticated layout support for languages which require it.

-----------

Not directly related, but see also this fork which uses SDF for text rendering: https://github.com/ocornut/imgui/pull/4056

**Comment by dgregorius** [comment]
My understanding (and I just started reading about it) is that the hinting algorithms used in ClearType (and it successor) are patented and therefore disabled by default (if implemented at all) in FreeType. I am just wondering if using font atlas is the right solution in the first place. It seems to me that ideally any GUI system would use the underlying OS text rendering for best results. In Qt you can use DirectWrite and Steam uses it actually. You can try it in the settings and it makes quite a difference if not enabled. I think Chrome also uses DirectWrite on Windows.

The SDF rendering is nice. But afaik it does not help with small text.

**Reply by PathogenDavid** [reply]
> It seems to me that ideally any GUI system would use the underlying OS text rendering for best results.

I don't necessarily disagree, but I don't think it's ever been a goal for Dear ImGui to provide a native OS feel.

Of you want to try you can override `ImFontAtlas::FontBuilderIO` with your own font atlas generation (this requires poking at `imgui_internal.h`, and do know that this interface will definitely change whenever dynamic font atlases are implemented.)

> I am just wondering if using font atlas is the right solution in the first place.

I don't think the font atlas is to blame here. Firefox uses an atlas too and they use the operating system's text rendering.

**Reply by thedmd** [reply]
To deal with one misconception. ImGui first is debug tool, crafted to integrate easily and minimize overhead in engine/app. Using Font Atlas is most efficient way to deal text rendering in this context.

## Support for subpixel rendering via atlas
It is possible to use ClearType-like rendering in ImGui:
* As you mentioned FreeType need to be build with `FT_CONFIG_OPTION_SUBPIXEL_RENDERING` enabled
* `imgui_freetype` then can be extended to pass [`FT_PIXEL_MODE_LCD`](https://freetype.org/freetype2/docs/reference/ft2-basic_types.html#ft_pixel_mode_lcd) or [`FT_PIXEL_MODE_LCD_V`](https://freetype.org/freetype2/docs/reference/ft2-basic_types.html#ft_pixel_mode_lcd_v)
* New flags can be passed to rasterizer through `ImFontConfig::FontBuilderFlags`

As an alternative to FreeType it is possible to write custom font rasterizer that use DirectWrite or GDI to rasterize glyphs on atlas.

## Support for custom text rendering 

Plugging DirectWrite into rendering pipeline is more challenging and will require modification of at least `ImFont` (`CalcTextSizeA`, `CalcWordWrapPositionA`, `RenderChar`, `RenderText`). Render functions can add custom draw calls, to be handled by custom backend on top of DirectWrite. DirectWrite is slow and will require caching to achieve good performance. As usual such task is leaved as an exercise for the user of the library. :)

Making such modification will be a prototype and probably suitable to not keep development of an app behind.
This also can serve as a basis for feature for ImGui to support this without need to modify library.


## ClearType and parents
ClearType was troubled by patents, but is no longer:
> Since August 2019, all patents related to ClearType color filtering have expired worldwide.

Source: https://freetype.org/patents.html

---

Did I managed to answer your question in full? :)

**Comment by dgregorius** [comment]
Great, thanks a lot. It seems the ClearType has been dropped for some other method on more recent hardware. I forgot the name. Something with monochromatic anti-aliasing iirc. If you happen to know what I mean, can this be enabled in FT as well?

**Comment by ocornut** [comment]
Note this PR for subpixel rendering:
https://github.com/ocornut/imgui/pull/2986
